common/lib/xmodule/xmodule/course_module.py

```diff
@@ -179,6 +179,156 @@ def to_json(self, values):
+
+        value = self._get_proctoring_value(value, proctoring_backend_settings)
+
+        return value
+
+    def _get_proctoring_value(self, value, proctoring_backend_settings):
+        """
+        Return a proctoring value that includes any inherited attributes from the platform defaults
+        for the backend or rules.
+        """
+        proctoring_provider = value.get('backend', None)
+        proctoring_provider_rules = value.get('rules', None)
+
+        # if both are missing from the value, return the default
+        if proctoring_provider is None and proctoring_provider_rules is None:
+            return self.default
+
+        # if provider is missing, but rules are not, use the default provider
+        # TODO: this would cause a mismatch between the provider and the rules...
+        if proctoring_provider is None and proctoring_provider_rules is not None:
+            value['backend'] = proctoring_backend_settings.get('DEFAULT', None)
+
+        # if rules are missing, but provider is not, use the default rules for the provider
+        if proctoring_provider is not None and proctoring_provider_rules is None:
+            value['rules'] = proctoring_backend_settings.get(value['backend'], {}).get('default_rules', {})
+
+        proctoring_provider_rules_set = (
+            proctoring_backend_settings
+            .get(proctoring_provider, {})
+            .get('default_rules', {})
+        )
+        proctoring_provider_rules = value.get('rules', None)
+
+        # add back in any missing rules for the provider
+        for default_rule, is_enabled in iteritems(proctoring_provider_rules_set):
+            if default_rule not in proctoring_provider_rules:
+                    proctoring_provider_rules[default_rule] = is_enabled
+        return value
+
+    def _validate_proctoring_backend(self, value, proctoring_backend_settings):
+        """
+        Validate the value for the proctoring backend. If the proctoring backend value is
+        specified, and it is not one of the backends configured at the platform level, return
+        a list of error messages to the caller.
+        """
+        errors = []
+
+        proctoring_provider_whitelist = [provider for provider in proctoring_backend_settings if provider != 'DEFAULT']
+        proctoring_provider_whitelist.sort()
+        proctoring_provider = value.get('backend', None)
+
+        if proctoring_provider and proctoring_provider not in proctoring_provider_whitelist:
+            errors.append(
+                _('The selected proctoring backend, {proctoring_backend}, is not a valid backend. '
+                    'Please select from one of {available_backends}.')
+                .format(
+                    proctoring_backend=proctoring_provider,
+                    available_backends=proctoring_provider_whitelist
+                )
+            )
+
+        return errors
+
+    def _validate_proctoring_rules(self, value, proctoring_backend_settings):
```

Comment:
There's some duplication in these two methods. It seems like it could just be one method, considering they're never called separately.